<h1> TITANIC PROJECT </h1>

#### Người thực hiện: Trần Minh Tiến

Các công đoạn tiến hành:
- Hiểu dữ liệu
- Xử lý dữ liệu đầu vào
- Áp dụng mô hình vào dữ liệu

#### Thêm các thư viện cần thiết

In [128]:
import pandas as pd
import numpy as np

Đọc dữ liệu từ file

In [129]:
data_train = pd.read_csv("./train.csv")
data_test = pd.read_csv("./test.csv")

Kiểm tra dữ liệu vừa đọc

In [130]:
data_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [131]:
data_test.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

Chọn những cột cần thiết cho training đồng thời ta tách cột kết quả dự đoán thành dataframe `label`

In [132]:
cols = ['PassengerId','Sex','Age','SibSp','Parch','Fare','Embarked']
df_train , df_test = data_train[cols],data_test[cols]
label = data_train['Survived']

Kiểm tra dữ liệu

In [133]:
df_train.head()

PassengerId     Sex   Age  SibSp  Parch     Fare Embarked
0            1    male  22.0      1      0   7.2500        S
1            2  female  38.0      1      0  71.2833        C
2            3  female  26.0      0      0   7.9250        S
3            4  female  35.0      1      0  53.1000        S
4            5    male  35.0      0      0   8.0500        S

In [134]:
df_train.shape

(891, 7)

In [135]:
df_train.isnull().sum()

PassengerId      0
Sex              0
Age            177
SibSp            0
Parch            0
Fare             0
Embarked         2
dtype: int64

In [137]:
df_test.isnull().sum()

PassengerId     0
Sex             0
Age            86
SibSp           0
Parch           0
Fare            1
Embarked        0
dtype: int64

#### Kiểm tra dữ liệu ta thấy:
- DataFrame `df_train` , cột Age có 177 giá trị missing-value và Embarked có 2 giá trị missing-value
- DataFrame `df_test` , cột Age có 86 giá trị missing-value và Fare có 1 giá trị missing-value

Xử lý missing value cột fare của dataframe test  bằng cách thay thế missing value bằng trung bình của cột fare

In [138]:
df_test['Fare'].fillna(np.mean(df_test['Fare']),inplace=True)

C:\Users\asus1\anaconda3\lib\site-packages\pandas\core\generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


Kiểm tra các cột bị missing value trong data frame `df_train`

In [139]:
df_train[df_train['Embarked'].isnull()]

PassengerId     Sex   Age  SibSp  Parch  Fare Embarked
61            62  female  38.0      0      0  80.0      NaN
829          830  female  62.0      0      0  80.0      NaN

Dòng 61 và 829 đang có dữ liệu NaN , ta loại bỏ các dòng này trong dữ liệu đồng thời loại bỏ luôn trong file label để đồng nhất dữ liệu

In [140]:
df_train.drop([61,829],axis = 0,inplace = True)
label.drop([61,829],axis = 0,inplace = True)

C:\Users\asus1\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [141]:
df_train.isnull().sum()

PassengerId      0
Sex              0
Age            177
SibSp            0
Parch            0
Fare             0
Embarked         0
dtype: int64

Thay thế dữ liệu cột Age trong df_train và df_test bằng giá trị trung bình  của từng cột

In [142]:
age_mean_train = df_train['Age'].mean()
age_mean_test = df_test['Age'].mean()

In [143]:
df_train.loc[:,'Age'].fillna(age_mean_train,inplace = True)
df_test.loc[:,'Age'].fillna(age_mean_test,inplace = True)

C:\Users\asus1\anaconda3\lib\site-packages\pandas\core\generic.py:6392: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


In [144]:
df_train.isnull().sum()

PassengerId    0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64

Kiểm tra cột `Embarked` ta thấy cột này chứa dữ liệu loại category có 3 giá trị `S` , `C` và `Q`

In [145]:
df_train['Embarked'].unique()

array(['S', 'C', 'Q'], dtype=object)

In [146]:
df_test['Embarked'].unique()

array(['Q', 'S', 'C'], dtype=object)

Sử dụng thư viện sklearn hỗ trợ việc gán Label cho cột `Embarked` , và `Sex`

In [147]:
from sklearn.preprocessing import LabelEncoder

In [148]:
le = LabelEncoder()
le.fit(df_train['Embarked'])
df_train.loc[:,'Embarked'] = le.transform(df_train['Embarked']).copy()
df_test.loc[:,'Embarked']= le.transform(df_test['Embarked']).copy()

C:\Users\asus1\anaconda3\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


Xử lý dữ liệu cột `Sex`
- Male = 0 
- Female = 1

In [149]:
le.fit(df_train['Sex'])
df_train.loc[:,'Sex'] = le.transform(df_train['Sex']).copy()
df_test.loc[:,'Sex']= le.transform(df_test['Sex']).copy()

In [150]:
df_train.head()

PassengerId  Sex   Age  SibSp  Parch     Fare  Embarked
0            1    1  22.0      1      0   7.2500         2
1            2    0  38.0      1      0  71.2833         0
2            3    0  26.0      0      0   7.9250         2
3            4    0  35.0      1      0  53.1000         2
4            5    1  35.0      0      0   8.0500         2

Kiểm tra dữ liệu ta thấy rằng dữ liệu trong `Embarked` đã chuyển thành số tương ứng, ta giải quyết xong việc chuyển các dữ liệu dạng text sang số để thuận tiện cho việc training Model , tuy nhiên dữ liệu có phạm vị giá trị khá lớn (Cột `Age` và `Fare`) nên ta cần `Chuẩn hoá dữ liệu` về cùng một khoảng giá trị.

Sử dụng phương pháp Standard Scaler (giống việc đưa dữ liệu về phân phối chuẩn)

Công thức: 
- *(x - mean)/std*

Trong đó: 
- x là giá trị từng dòng
- mean là giá trị trung bình cột
- std là độ lệch chuẩn cùa cột

In [151]:
def StdScaler(x,mean,std):
    return (x - mean)/std

Tính các giá trị mean và std của từng dữ liệu

In [152]:
age_df_train_std = np.std(df_train.loc[:,'Age'])
age_df_train_mean = np.mean(df_train.loc[:,'Age'])
age_df_test_std = np.std(df_test.loc[:,'Age'])
age_df_test_mean = np.mean(df_test.loc[:,'Age'])

In [153]:
df_train.loc[:,'Age'] = df_train.loc[:,'Age'].apply(lambda x: StdScaler(x,age_df_train_mean,age_df_train_std)).copy()
df_test.loc[:,'Age'] = df_test.loc[:,'Age'].apply(lambda x: StdScaler(x,age_df_test_mean,age_df_test_std)).copy()

C:\Users\asus1\anaconda3\lib\site-packages\pandas\core\indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


In [154]:
fare_df_train_std = np.std(df_train.loc[:,'Fare'])
fare_df_train_mean = np.mean(df_train.loc[:,'Fare'])

fare_df_test_std = np.std(df_test.loc[:,'Fare'])
fare_df_test_mean = np.mean(df_test.loc[:,'Fare'])

In [155]:
df_train.loc[:,'Fare'] = df_train.loc[:,'Fare'].apply(lambda x: StdScaler(x,fare_df_train_mean,fare_df_train_std)).copy()
df_test.loc[:,'Fare'] =df_test.loc[:,'Fare'].apply(lambda x: StdScaler(x,fare_df_test_mean,fare_df_test_std)).copy()

Kiểm tra dữ liệu đã chuẩn hoá

In [156]:
df_train.head()

PassengerId  Sex       Age  SibSp  Parch      Fare  Embarked
0            1    1 -0.589620      1      0 -0.500240         2
1            2    0  0.644848      1      0  0.788947         0
2            3    0 -0.281003      0      0 -0.486650         2
3            4    0  0.413385      1      0  0.422861         2
4            5    1  0.413385      0      0 -0.484133         2

In [157]:
df_test.head()

PassengerId  Sex       Age  SibSp  Parch      Fare  Embarked
0          892    1  0.334993      0      0 -0.498407         1
1          893    0  1.325530      1      0 -0.513274         2
2          894    1  2.514175      0      0 -0.465088         1
3          895    1 -0.259330      0      0 -0.483466         2
4          896    0 -0.655545      1      1 -0.418471         2

### Split dataset

Vì chúng ta chỉ có một dataset có dữ liệu chính xác (được cho trước) nên ta chia dữ liệu thành 2 phần :
- 1 phần để training dữ liệu
- 1 phần cho việc test dữ liệu của model vừa được train

Để thuận tiện sklearn có hỗ trợ module cho việc phân chia dữ liệu

In [158]:
from sklearn.model_selection import train_test_split

In [159]:
x_train,x_test,y_train,y_test = train_test_split(df_train,label,random_state = 40 , test_size = 0.3)

- `Random_state` có tác dụng để giữ ổn định dữ liệu khi chạy lại đoạn code
- `Test_size` là số lượng dòng được dùng cho dữ liệu test (ở đây 0.3 là 30%)
- `x_train` là dữ liệu train , `y_train` là đầu ra tương ứng dữ liệu của x_train
- `x_test` là dữ liệu test , `y_test` là đầu ra tương ứng dữ liệu của x_test

Dữ liệu được chia ra ở x_test và x_train sẽ tương tứng 3 phần cho test và 7 phần cho train như đã set ở test_size

## Training model 

Dữ liệu đã được chuẩn bị bao gồm `df_train` và `df_test` , `df_test` là dữ liệu mà bài toán đặt ra model chúng ta cần đoán được , muốn được vậy dữ liệu training của chúng ta phải thực hiện được tốt trên tập `df_train` (dữ liệu thực tế) . Việc chia `df_train` thành tập train và test cũng có tác dụng tương tự , việc chuẩn đoán mô hình vừa sử dụng chạy đúng thực tế hay không đòi hỏi phải có một tập dữ liệu chính xác , tuy nhiên chúng ta chỉ có một tập `df_train` nên cách duy nhất là chia tập trên thành 2 phần.

Việc đầu tiên ta cần thực hiện là cho model chạy gần chính xác trên tập `df_train` sau khi có kết quả gần hài lòng thì ta thực thi trên tập dữ liệu đầu đề `df_test`.

Có thể có nhiều thuật toán xử lý dữ liệu . tuy nhiên mỗi thuật toán cho một độ chính xác đầu ra khác nhau , để kiểm tra dữ liệu dự đoán bằng model và dữ liệu thật , ta sử dụng một module có sẵn trong `sklearn` là `accuracy_score`.

In [160]:
from sklearn.metrics import accuracy_score

#### Ta thử với thuật toán đầu tiên là Passive Aggressive Classifier trong thư viện sklearn

In [177]:
from sklearn.linear_model import PassiveAggressiveClassifier

Các thao tác sử dụng thư viện ML có thể xem trên youtube Code Explore

In [178]:
pac = PassiveAggressiveClassifier()
pac.fit(x_train,y_train)

PassiveAggressiveClassifier()

Ta tiến hành chuẩn đoán dữ liệu đầu ra bằng thuật toán Passive Aggressive Class trên tập dữ liệu x_test

In [179]:
y_pac_pred = pac.predict(x_test)

In [180]:
y_pac_pred

array([0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0], dtype=int64)

Ta so sánh dữ liệu vừa dữ đoán và dữ liệu thực tế trong tập `y_test`

In [181]:
pd.DataFrame(y_pac_pred,y_test)

0
Survived   
0         0
0         1
1         0
0         0
0         0
...      ..
0         0
0         0
0         0
0         0
0         0

[267 rows x 1 columns]

Ta thấy có sự khá tương đồng giữa 2 cột dữ liệu , điều ta mong muốn là dữ liệu ta đoán bằng model phải gần chính xác nhất so với dữ liệu thực `y_test` , tuy nhiên nhìn mắt thường rất khó để xác định được việc giống nhau là như thế nào nên ta sử dụng module `accuracy_score` để máy thực hiện so khớp

In [182]:
accuracy_score(y_test,y_pac_pred)

0.797752808988764

Kết quả trả về là 79% (mỗi lần chạy có thể khác nhau vì predict khác nhau) , tuy nhiên ta vẫn muốn tỷ lệ đó cao hơn . Để việc tỷ lệ đó cao hơn thì ta có thể xử lý lại data ban đầu hoặc thay đổi về thuật toán ML đang sử dụng

#### Thuật toán Support Vector Machine trong sklearn

Ta thực hiện tương tự các bước ở trên

In [183]:
from sklearn.svm import SVC

SVM_KERNEL = "linear"
SVM_C = 10
SVM_GAMMA = 0.00001

svm_model = SVC(kernel = SVM_KERNEL, C = SVM_C, gamma = SVM_GAMMA)
svm_model.fit(x_train,y_train)
y_test_pred_2 = svm_model.predict(x_test)

In [184]:
accuracy_score(y_test_pred_2,y_test)

0.7940074906367042

#### Áp dụng model vào dữ liệu đầu đề 

Trong vài trường hợp SVC trả về kết quả tốt hơn nên ta sử dụng thuật toán này cho dữ liệu đầu đề đưa ra `df_test`

In [52]:
y_pred = svm_model.predict(df_test)
y_pred

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,

Ta không được cho kết quả đầu ra nên không thể kiểm tra accuracy , việc xác định accuracy sẽ quyết định số điểm trên Kaggle

Xuất file kết quả theo đúng định dạng file `submission.csv` được cho để làm mẫu

In [186]:
result = pd.DataFrame(df_test.loc[:,'PassengerId'])

In [187]:
result['Survived'] = y_pred

In [188]:
result.head()

PassengerId  Survived
0          892         0
1          893         0
2          894         0
3          895         0
4          896         1

In [ ]:
result.to_csv("out.csv", index=False)

Tiến hành nộp file `out.csv` lên Kaggle để xác định accuracy